File: Go/countdown.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/ernestosuarez/itertools"
	"github.com/karrick/gorpn"
)

var equations [][]string
var sub = make(map[string]float64)

func main() {
	start := time.Now()
	var operators = []string{"+", "-", "*", "/"}
	var number = []string{"1", "1", "2", "2", "3", "3", "4", "4", "5", "5", "6", "6", "7", "7", "8", "8", "9", "9", "10", "10", "25", "50", "75", "100"}
	var ops_needed int = -1
	var sol []string

	// Creates a dict of all 3 digit numbers. Values set to 0
	threeDigits := make(map[int]int)
	for i := 101; i < 1000; i++ {
		threeDigits[i] = 0
	}

	// Deals with duplicates for Combinations
	var cNums [][]string
	allKeys := make(map[string]bool)
	comb := itertools.CombinationsStr(number, 5)
	for item := range comb {
		var strSlc string = strings.Join(item[:], ",") // Turns slice into string to create unique key
		if _, value := allKeys[strSlc]; !value {
			allKeys[strSlc] = true
			cNums = append(cNums, [][]string{item}...)
		}
	}

	// Deals with duplicates for Permutations
	for _, item := range cNums {
		var pNums [][]string
		permKeys := make(map[string]bool)
		perm := itertools.PermutationsStr(item, len(item))
		for x := range perm {
			var permSlc string = strings.Join(x[:], ",") // Turns slice into string to create unique key
			if _, value := permKeys[permSlc]; !value {
				permKeys[permSlc] = true
				pNums = append(pNums, [][]string{x}...)
			}
		}

		for _, item := range pNums {
			var current []string
			rpn(item, operators, current, ops_needed)
		}

		sol = append(sol, []string(equate(equations))...) // Add all (valid) solutions to a slice

		for _, v := range sol {
			i, _ := strconv.Atoi(v) // Convert solutions into ints
			threeDigits[i] += 1     // Increment the dict by one where a solution exists
		}
		sol = nil       // Reset sol so it doesn't kill the memory
		equations = nil // Reset equations for next round
	}

	// Write the key, value to a csv file
	file, _ := os.Create("GoNumbers.csv")
	defer file.Close()

	w := csv.NewWriter(file)
	defer w.Flush()

	var keys []int
	for k := range threeDigits {
		keys = append(keys, k)
	}
	sort.Ints(keys)
	for _, k := range keys {
		row := []string{strconv.Itoa(k), strconv.Itoa(threeDigits[k])}
		w.Write(row)
	}

	elapsed := time.Since(start)
	fmt.Printf("Elapsed: %v\n", elapsed)
}

func rpn(nums []string, ops []string, current []string, ops_needed int) {
	if ops_needed == 0 && len(nums) == 0 {
		c := make([]string, len(current))
		copy(c, current)
		equations = append(equations, [][]string{c}...)
	}

	if ops_needed > 0 {
		for _, op := range ops {
			current = append(current, op)
			rpn(nums, ops, current, ops_needed-1)
			current = current[:len(current)-1]
		}
	}

	if size := len(nums); size > 0 {
		v := nums[size-1]            // Save pop value
		nums = nums[:size-1]         // Pop last element
		current = append(current, v) // Push element
		rpn(nums, ops, current, ops_needed+1)
	}
}

func equate(equations [][]string) []string {
	var listResult []string
	var eqString string
	for _, item := range equations { // equations = [[1,2,+,3,+] [1,2,+,3-] etc.]
		var eq []string
		var temp []string
		for _, term := range item { // item = [1,2,+,3,+]
			_, err := strconv.Atoi(term) // Attempt to convert term to int
			if err == nil {
				eq = append(eq, term) // Append term to eq if int i.e. [1,2]
			} else {
				temp = append(temp, eq[len(eq)-2], eq[len(eq)-1], term) // Create postfix equation [1,2,+]
				eq = eq[:len(eq)-2]                                     // Remove last two elements from eq. Setting to nil will break everything

				if term == "+" || term == "*" { // Turn postfix equation into standardised string for key
					eqString = SortString(temp)
				} else {
					eqString = strings.Join(temp, ",")
				}

				if val, ok := sub[eqString]; ok { // If value for key exists, use the value. Huge save on comp. time
					v := fmt.Sprint(val)
					if isIntegral(val) && val > 0 {
						if val > 100 && val < 1000 {
							listResult = append(listResult, []string{v}...)
						}
						eq = append(eq, v)
					} else {
						break
					}
				} else { // If value doesn't exist, compute it and add it to dict.
					expression, _ := gorpn.New(eqString)
					result, _ := expression.Evaluate(nil)
					sub[eqString] = result
					rs := fmt.Sprint(result)
					if isIntegral(result) && result > 0 {
						if result > 100 && result < 1000 {
							listResult = append(listResult, []string{rs}...)
						}
						eq = append(eq, rs)
					} else {
						break
					}
				}
				temp = nil
			}
		}
	}
	return listResult
}

func isIntegral(val float64) bool {
	return val == float64(int(val))
}

func SortString(w []string) string {
	sort.Sort(sort.Reverse(sort.StringSlice(w)))
	return strings.Join(w, ",")
}

// func check(val float64) bool {
// 	if isIntegral(val) && val > 0 {
// 		if val > 100 && val < 1000 {
// 			return true
// 		}
// 	} else {
// 		return false
// 	}
// }

File: Python/countdown.py
from itertools import permutations, combinations
import copy
import time

def rpn(nums: list, ops: list, current: list, ops_needed: int = -1):
    if ops_needed == 0 and len(nums) == 0:
        c = copy.deepcopy(current)
        equations.append(c)
    
    if ops_needed > 0:
        for op in ops:
            current.append(op)
            rpn(nums, ops, current, ops_needed-1)
            current.pop()
    
    if len(nums) > 0:
        var = nums.pop()
        current.append(var)
        rpn(nums, ops, current, ops_needed+1)
        current.pop()
        nums.append(var)

def calculate(equa: list):
    stack, eqString, test = [], '', {}
    for item in equa:
        for term in item:
            if type(term) == int:
                stack.append(term)
            else:
                temp = [stack.pop(-1), stack.pop(-1), term]
                c = [str(int) for int in temp]

                if term == '+' or term == '*':
                    c.sort()
                    eqString = ','.join(c)
                else:
                    eqString = ','.join(c)
                
                match temp:
                    case [1, _, '*']:
                        stack.append(temp[1])
                        continue
                    case [_, 1, '/'|'*']:
                        stack.append(temp[0])
                        continue

                if eqString in test.keys():
                    v = test[eqString]
                    if v:
                        stack.append(v)
                        continue
                    else:
                        break
                
                if eqString in sub.keys():
                    v = sub[eqString]
                    test[eqString] = v
                    if v:
                        if 100 < v < 1000:
                            dict[v] += 1
                        stack.append(v)
                    else:
                        break
                else:
                    exp = equate(temp[1],temp[0], term)
                    sub[eqString] = exp
                    test[eqString] = exp
                    if exp:
                        if 100 < exp < 1000:
                            dict[exp] += 1
                        stack.append(exp)
                    else:
                        break
                
        stack.clear()

def equate(a: int, b: int, term: str):
    c = 0
    match term:
        case '+':
            c = a+b
        case '*':
            c = a*b
        case '-':
            c = a-b
        case '/':
            c = a/b

    if (c%1) == 0 and c > 0:
        return int(c)
    else:
        return 0

def main():
    numbers = [1,1,2,2,3,3,4,4,5,5,6,6,7,7,8,8,9,9,10,10,25,50,75,100] # All the numbers in the Countdown rules
    operators = ['+', '-', '*', '/']
    cDupe = set()

    # Deals with duplicates for Combinations
    cNums = sorted(combinations(numbers, 4))
    for item in cNums:
        cDupe.add(item)
    cNums.clear()

    # Deals with duplicates for Permutations
    for item in cDupe:
        pNums = permutations(item)
        perm = set()
        for item in pNums:
            perm.add(item)
        
        for item in perm:
            rpn(list(item), operators, [])
        
        calculate(equations) # Equations gets filled in the rpn function

        equations.clear()

    with open('Python/numbers.csv', 'w') as file:
        for key in dict.keys():
            file.write(str(key) + "," + str(dict[key]) + '\n')

if __name__ == '__main__':
    equations = []
    dict = {}
    sub = {}

    # Creates a dict of all 3 digit numbers. Values set to 0
    keys = range(101, 1000, 1)
    for i in keys:
        dict[i] = 0

    start = time.perf_counter()
    main()
    finish = time.perf_counter()
    print(f'Elapsed: {round(finish-start,2)}s')